akka-stream/src/main/scala/akka/stream/impl/fusing/Interpreter.scala

```diff
@@ -513,67 +642,94 @@ private[akka] class OneBoundedInterpreter(ops: Seq[Stage[_, _]], val forkLimit:
 
   def isFinished: Boolean = pipeline(Upstream) == Finished && pipeline(Downstream) == Finished
 
-  /**
-   * This method injects a Context to each of the BoundaryStages. This will be the context returned by enter().
-   */
-  private def initBoundaries(): Unit = {
-    var op = 0
-    while (op < pipeline.length) {
-      // FIXME try to change this to a pattern match `case boundary: BoundaryStage`
-      // but that doesn't work with current Context types
-      if (pipeline(op).isInstanceOf[BoundaryStage]) {
-        pipeline(op).asInstanceOf[BoundaryStage].bctx = new State {
-          val entryPoint = op
-
-          override def run(): Unit = ()
-          override def advance(): Unit = ()
-
-          override def push(elem: Any): DownstreamDirective = {
-            ReactiveStreamsCompliance.requireNonNullElement(elem)
-            activeOpIndex = entryPoint
-            super.push(elem)
-            execute()
-            null
-          }
+  private class EntryState(name: String, position: Int) extends State {
+    val entryPoint = position
 
-          override def pull(): UpstreamDirective = {
-            activeOpIndex = entryPoint
-            super.pull()
-            execute()
-            null
-          }
+    override def run(): Unit = ()
+    override def advance(): Unit = ()
 
-          override def finish(): FreeDirective = {
-            activeOpIndex = entryPoint
-            super.finish()
-            execute()
-            null
-          }
+    override def enter(evt: Any): Unit = {
+      activeOpIndex = entryPoint
+      (currentOp: Any) match {
+        case a: AsyncStage[Any, Any, Any] ⇒
+          if (Debug) println("    " * entryPoint + "-ai- " + evt)
+          a.onAsyncInput(evt, this)
+        case other ⇒ throw new IllegalStateException("cannot enter() a non-AsyncStage")
+      }
+    }
 
-          override def fail(cause: Throwable): FreeDirective = {
-            activeOpIndex = entryPoint
-            super.fail(cause)
-            execute()
-            null
-          }
+    override def comingFromUpstream = false
+    override def comingFromDownstream = false
+    override def comingFromSide = true
 
-          override def hold(): FreeDirective = {
-            activeOpIndex = entryPoint
-            super.hold()
-            execute()
-            null
-          }
+    override def push(elem: Any): DownstreamDirective = {
+      ReactiveStreamsCompliance.requireNonNullElement(elem)
+      activeOpIndex = entryPoint
+      super.push(elem)
+      execute()
+      null
+    }
 
-          override def pushAndPull(elem: Any): FreeDirective = {
-            ReactiveStreamsCompliance.requireNonNullElement(elem)
-            activeOpIndex = entryPoint
-            super.pushAndPull(elem)
-            execute()
-            null
-          }
+    override def pull(): UpstreamDirective = {
+      activeOpIndex = entryPoint
```

Comment:
I don’t know what you mean by “closures”, but yes, this could be implemented differently by restructuring how the boundaries enter their stage/context. I’ll try it out.